rust/src/array.rs

```diff
@@ -18,447 +18,562 @@
+def_primitive_array!(DataType::Float32, f32);
+def_primitive_array!(DataType::Float64, f64);
+
+/// A list array where each element is a variable-sized sequence of values with the same
+/// type.
+pub struct ListArray {
+    data: ArrayDataRef,
+    values: ArrayRef,
+    value_offsets: RawPtrBox<i32>,
 }
 
-impl<T> PrimitiveArray<T>
-where
-    T: ArrowPrimitiveType,
-{
-    pub fn new(data: Buffer<T>, null_count: usize, validity_bitmap: Option<Bitmap>) -> Self {
-        PrimitiveArray {
-            len: data.len(),
-            data: data,
-            null_count,
-            validity_bitmap,
-        }
+impl ListArray {
+    /// Returns an reference to the values of this list.
+    pub fn values(&self) -> ArrayRef {
+        self.values.clone()
     }
 
-    pub fn len(&self) -> usize {
-        self.len
+    /// Returns the value type of this list.
+    pub fn value_type(&self) -> DataType {
+        self.values.data().data_type()
     }
 
-    pub fn get(&self, i: usize) -> &T {
-        self.data.get(i)
+    /// Returns the offset for value at index `i`.
+    #[inline]
+    pub fn value_offset(&self, i: i64) -> i32 {
+        self.value_offset_at(self.data.offset() + i)
     }
 
-    pub fn iter(&self) -> BufferIterator<T> {
-        self.data.iter()
+    /// Returns the length for value at index `i`.
+    #[inline]
+    pub fn value_length(&self, mut i: i64) -> i32 {
+        i += self.data.offset();
+        self.value_offset_at(i + 1) - self.value_offset_at(i)
     }
 
-    pub fn buffer(&self) -> &Buffer<T> {
-        &self.data
+    #[inline]
+    fn value_offset_at(&self, i: i64) -> i32 {
+        unsafe { *self.value_offsets.get().offset(i as isize) }
     }
+}
 
-    /// Determine the minimum value in the array
-    pub fn min(&self) -> Option<T> {
-        let mut n: Option<T> = None;
-        match &self.validity_bitmap {
-            &Some(ref bitmap) => for i in 0..self.len {
-                if bitmap.is_set(i) {
-                    let mut m = self.data.get(i);
-                    match n {
-                        None => n = Some(*m),
-                        Some(nn) => if *m < nn {
-                            n = Some(*m)
-                        },
-                    }
-                }
-            },
-            &None => for i in 0..self.len {
-                let mut m = self.data.get(i);
-                match n {
-                    None => n = Some(*m),
-                    Some(nn) => if *m < nn {
-                        n = Some(*m)
-                    },
-                }
-            },
+/// Constructs a `ListArray` from an array data reference.
+impl From<ArrayDataRef> for ListArray {
+    fn from(data: ArrayDataRef) -> Self {
+        debug_assert!(data.buffers().len() == 1);
+        debug_assert!(data.child_data().len() == 1);
+        let values = make_array(data.child_data()[0].clone());
+        let value_offsets = data.buffers()[0].raw_data() as *const i32;
```

Comment:
not sure if this pointer here is actually aligned and points to a valid location?